### PR TITLE
Add another explicit page zeroing when MADV_DONTNEED is not enough

### PR DESCRIPTION
**Pull request by jedisct1**
Needless to say that clearing the whole heap has some performance implications :(

### CODE REVIEW

**Comment by jedisct1** [comment]
Do we really need to clear the whole heap?

Can we have non-zeroed pages after `alloc.heap_accessible_size`?

**Comment by acfoltzer** [comment]
We don't ever shrink the heap, so I think we only need to zero up until the accessible size.

You'll also likely need to add similar code in the implementation of `reset`.

I'd also appreciate if you could keep that commented `eprintln!` around; it's useful to have for debugging things like this.

**Comment by jedisct1** [comment]
Thanks Adam. The commented out debugging statement has been re-added.

Explicit zeroing was already done in `reset()`.

File: lucet-runtime/lucet-runtime-internals/src/region/mmap.rs [hunk 1]
```diff
@@ -100,6 +101,13 @@ impl RegionInternal for MmapRegion {
         {
             // eprintln!("setting none {:p}[{:x}]", *ptr, len);
             unsafe {
+                // MADV_DONTNEED is not guaranteed to clear pages on non-Linux systems
+                #[cfg(not(target_os = "linux"))]
+                {
+                    mprotect(*ptr, *len, ProtFlags::PROT_READ | ProtFlags::PROT_WRITE)
+                        .expect("mprotect succeeds during drop");
+                    memset(*ptr, 0, *len);
```
Comment (acfoltzer) [inline lucet-runtime/lucet-runtime-internals/src/region/mmap.rs]: should we set `PROT_NONE` again after this `memset`?

Comment (awortman-fastly) [reply lucet-runtime/lucet-runtime-internals/src/region/mmap.rs]: that's unconditionally done a few lines down from here :)

Comment (acfoltzer) [reply lucet-runtime/lucet-runtime-internals/src/region/mmap.rs]: so it is, lol :+1:

**Pull request merged**